compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -934,6 +957,54 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
             .collect();
         let mut fields = fields?;
 
+        if let AggregateTy::Def(_, _) = &mut ty
+            && let Some(&first_field) = fields.first()
+            && let Value::Projection(copy_from_value, _) = *self.get(first_field)
+            && fields.iter().enumerate().all(|(index, &v)| {
+                // All fields must correspond one-to-one and come from the same aggregate value.
+                if let Value::Projection(pointer, ProjectionElem::Field(from_index, _)) =
+                    *self.get(v)
+                    && copy_from_value == pointer
+                    && from_index.index() == index
+                {
+                    return true;
+                }
+                false
+            })
+            && let Some(mut copy_from_local_value) = if let Value::Projection(pointer, proj) =
+                *self.get(copy_from_value)
+                && let Some(ProjectionElem::Downcast(_, read_variant)) =
+                    self.try_as_place_elem(proj, location)
+            {
+                // The copied variant must be identical.
+                if variant_index != read_variant { None } else { Some(pointer) }
+            } else {
+                Some(copy_from_value)
+            }
+        {
+            let mut projection = SmallVec::<[PlaceElem<'tcx>; 1]>::new();
+            loop {
+                if let Some(local) = self.try_as_local(copy_from_local_value, location) {
+                    projection.reverse();
+                    let place =
+                        Place { local, projection: tcx.mk_place_elems(projection.as_slice()) };
+                    if rvalue_ty == place.ty(self.local_decls, tcx).ty {
+                        self.reused_locals.insert(local);
+                        *rvalue = Rvalue::Use(Operand::Copy(place));
+                        return Some(copy_from_value);
+                    }
+                    break;
+                } else if let Value::Projection(pointer, proj) = *self.get(copy_from_local_value)
+                    && let Some(proj) = self.try_as_place_elem(proj, location)
+                {
+                    projection.push(proj);
+                    copy_from_local_value = pointer;
+                } else {
+                    break;
+                }
+            }
+        }
+
```

Comment:
Should it be put in a standalone method? To use `?` and early returns and make it a bit more readable.

Reply:
Updated.